server/src/main/java/io/crate/metadata/cluster/AlterTableClusterStateExecutor.java

```diff
@@ -153,149 +150,49 @@ protected ClusterState execute(ClusterState currentState, AlterTableRequest requ
-        //           not_null : [],
-        //       },
-        //       indices: {
-        //          <index_name>: <options>
-        //      }
-        //       check_constraints: {
-        //           <constraint_name>: <expression>
-        //       }
-        //   }
-        //
-        //   - partitioned_by cannot be changed
-        //
-        //   - primary_keys are always additive
-        //      - If present in delta, ALL keys are present (We should change this to gain atomicity)
-        //
-        //   - not_null constraints are always additive
-        //      - If present in delta, NEW columns are present
-        //
-        //   - indices can only be added
-        //      - If present in delta, NEW indices are present
-        //
-        //   - check_constraints can be added OR removed
-        //      If removed, delta contains *ALL* but the one to remove
-        //      If added:
-        //          Empty if new column doesn't have a constraint
-        //          All constraints present if new column has a constraint
-        //          (We should change this to gain atomicity)
-        //
-        // Why is everything behaving differently? Because reasons 🤷
-        // This would be simpler if we distinguished between additions and deletions in the AlterTableRequest
-        Map<String, Object> mappingDeltaAsMap = request.mappingDeltaAsMap();
-        Map<String, Object> metaDelta = (Map<String, Object>) mappingDeltaAsMap.get("_meta");
-        if (metaDelta != null) {
-            var curPartitionedBy = currentMeta.get("partitioned_by");
-            if (curPartitionedBy != null) {
-                metaDelta.put("partitioned_by", curPartitionedBy);
-            }
-            var curPrimaryKeys = currentMeta.get("primary_keys");
-            if (curPrimaryKeys != null) {
-                metaDelta.putIfAbsent("primary_keys", curPrimaryKeys);
-            }
 
-            Map<String, Object> checkConstraints = (Map<String, Object>) metaDelta.get("check_constraints");
-            if (checkConstraints == null || checkConstraints.isEmpty()) {
-                var curCheckConstraints = currentMeta.get("check_constraints");
-                if (curCheckConstraints != null) {
-                    metaDelta.put("check_constraints", curCheckConstraints);
-                }
+            Map<String, Object> indexMapping = indexMetadata.mapping().sourceAsMap();
+            if (indexMapping.get(ES_MAPPING_NAME).equals(mappingDelta.get(ES_MAPPING_NAME))) {
+                // If mapping before an update and after the update is same,
+                // MapperService throws an error on the metadata update attempt.
+                // Since mappingDelta is a map with only 1 key-value ("dynamic" -> specified value)
+                // we can change 1 property in the mapping.
+                // We check previous version of the column policy to make sure that new value is a new one and we won't fall to the error.
```

Comment:
Maybe lets remove this comment at all as it should be clear that we do not want to modify the cluster for no reason. Also the mapper service won't throw an error but an assertion instead (assertions are disabled in the distribution, thus this mostly affect test runs). My question was only related to that I wasn't  reading the code carefully enough ;)

Reply:
removed comment and rebased